Mathlib/Analysis/Analytic/Constructions.lean

```diff
@@ -0,0 +1,323 @@
+    (𝕜 𝕝 : Type*) [NontriviallyNormedField 𝕜] [NontriviallyNormedField 𝕝] [NormedAlgebra 𝕜 𝕝] :
+    HasFPowerSeriesOnBall (fun x : 𝕝 ↦ (1 - x)⁻¹) (formalMultilinearSeries_geometric 𝕜 𝕝) 0 1 := by
+  constructor
+  · exact le_of_eq (formalMultilinearSeries_geometric_radius 𝕜 𝕝).symm
+  · exact one_pos
+  · intro y hy
+    simp_rw [zero_add, formalMultilinearSeries_geometric,
+        ContinuousMultilinearMap.mkPiAlgebraFin_apply,
+        List.prod_ofFn, Finset.prod_const,
+        Finset.card_univ, Fintype.card_fin]
+    apply hasSum_geometric_of_norm_lt_1
+    simpa only [←ofReal_one, Metric.emetric_ball, Metric.ball,
+      dist_eq_norm, sub_zero] using hy
+
+lemma analyticAt_inv_one_sub (𝕝 : Type*) [NontriviallyNormedField 𝕝] [NormedAlgebra 𝕜 𝕝] :
+    AnalyticAt 𝕜 (fun x : 𝕝 ↦ (1 - x)⁻¹) 0 :=
+  ⟨_, ⟨_, hasFPowerSeriesOnBall_inv_one_sub 𝕜 𝕝⟩⟩
+
+/-- If `𝕝` is a normed field extension of `𝕜`, then the inverse map `𝕝 → 𝕝` is `𝕜`-analytic
+away from 0. -/
+lemma analyticAt_inv {z : 𝕝} (hz : z ≠ 0) : AnalyticAt 𝕜 Inv.inv z := by
+  let f1 : 𝕝 → 𝕝 := fun a ↦ 1 / z * a
+  let f2 : 𝕝 → 𝕝 := fun b ↦ (1 - b)⁻¹
+  let f3 : 𝕝 → 𝕝 := fun c ↦ 1 - c / z
+  have feq : f1 ∘ f2 ∘ f3 = Inv.inv
+  · ext1 x
+    dsimp only [Function.comp_apply]
+    field_simp
+  have f3val : f3 z = 0 := by simp only [div_self hz, sub_self]
+  have f3an : AnalyticAt 𝕜 f3 z
+  · apply analyticAt_const.sub
+    simpa only [div_eq_inv_mul] using analyticAt_const.mul (analyticAt_id 𝕜 z)
+  exact feq ▸ (analyticAt_const.mul (analyticAt_id _ _)).comp
+    ((f3val.symm ▸ analyticAt_inv_one_sub 𝕝).comp f3an)
+
+/-- `x⁻¹` is analytic away from zero -/
+lemma analyticOn_inv : AnalyticOn 𝕜 (fun z ↦ z⁻¹) {z : 𝕝 | z ≠ 0} := by
+  intro z m; exact analyticAt_inv m
+
+/-- `(f x)⁻¹` is analytic away from `f x = 0` -/
+theorem AnalyticAt.inv {f : E → 𝕝} {x : E} (fa : AnalyticAt 𝕜 f x) (f0 : f x ≠ 0) :
+    AnalyticAt 𝕜 (fun x ↦ (f x)⁻¹) x :=
+  (analyticAt_inv f0).comp fa
+
+/-- `x⁻¹` is analytic away from zero -/
+theorem AnalyticOn.inv {f : E → 𝕝} {s : Set E} (fa : AnalyticOn 𝕜 f s) (f0 : ∀ x, x ∈ s → f x ≠ 0) :
+    AnalyticOn 𝕜 (fun x ↦ (f x)⁻¹) s :=
+  fun x m ↦ (fa x m).inv (f0 x m)
+
+/-- `f x / g x` is analytic away from `g x = 0` -/
+theorem AnalyticAt.div {f g : E → 𝕝} {x : E}
+    (fa : AnalyticAt 𝕜 f x) (ga : AnalyticAt 𝕜 g x) (g0 : g x ≠ 0) :
+    AnalyticAt 𝕜 (fun x ↦ f x / g x) x := by
+  simp_rw [div_eq_mul_inv]; exact fa.mul (ga.inv g0)
+
+/-- `f x / g x` is analytic away from `g x = 0` -/
+theorem AnalyticOn.div {f g : E → 𝕝} {s : Set E}
+    (fa : AnalyticOn 𝕜 f s) (ga : AnalyticOn 𝕜 g s) (g0 : ∀ x, x ∈ s → g x ≠ 0) :
+    AnalyticOn 𝕜 (fun x ↦ f x / g x) s := fun x m ↦
+  (fa x m).div (ga x m) (g0 x m)
+
+/-!
+### Finite sums and products of analytic functions
+-/
+
+/-- Finite sums of analytic functions are analytic -/
+theorem Finset.analyticAt_sum {f : α → E → F} {c : E}
+    (N : Finset α) (h : ∀ n, n ∈ N → AnalyticAt 𝕜 (f n) c) :
```

Comment:
```suggestion
    (N : Finset α) (h : ∀ n ∈ N, AnalyticAt 𝕜 (f n) c) :
```
It's exactly the same thing, just more readable. Same thing below.

Reply:
Done.